dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/expression/dataitem/ItemYearlyPeriodCount.java

```diff
@@ -0,0 +1,118 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.expression.dataitem;
+
+import static java.lang.String.valueOf;
+import static org.hisp.dhis.calendar.DateTimeUnit.fromJdkDate;
+import static org.hisp.dhis.parser.expression.ParserUtils.DOUBLE_VALUE_IF_NULL;
+import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;
+import static org.hisp.dhis.period.PeriodType.getPeriodFromIsoString;
+
+import java.util.List;
+
+import org.hisp.dhis.calendar.DateTimeUnit;
+import org.hisp.dhis.parser.expression.CommonExpressionVisitor;
+import org.hisp.dhis.parser.expression.ExpressionItem;
+import org.hisp.dhis.period.BiWeeklyAbstractPeriodType;
+import org.hisp.dhis.period.CalendarPeriodType;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodType;
+import org.hisp.dhis.period.WeeklyAbstractPeriodType;
+
+/**
+ * Expression item [yearlyPeriodCount]
+ *
+ * @author Jim Grace
+ */
+public class ItemYearlyPeriodCount
+    implements ExpressionItem
+{
+    @Override
+    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        return DOUBLE_VALUE_IF_NULL;
+    }
+
+    @Override
+    public Double evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        List<Period> periods = visitor.getParams().getPeriods();
+
+        if ( periods.size() != 1 )
+        {
+            return 0.0; // Not applicable
+        }
+
+        Period period = periods.get( 0 );
+        PeriodType periodType = period.getPeriodType();
+
+        if ( periodType instanceof WeeklyAbstractPeriodType )
+        {
+            return weeksInYear( period );
+        }
+        else if ( periodType instanceof BiWeeklyAbstractPeriodType )
+        {
+            return biWeeksInYear( period );
+        }
+        else if ( periodType instanceof CalendarPeriodType && periodType.getFrequencyOrder() < 365 )
```

Comment:
For years that have 366 days, do we need some specific checks?

Reply:
Yes, thanks. The period count for 366 daily periods is handled below within `countPeriodsWithinThisYear`. I've replaced `< 365` to `< YearlyPeriodType.FREQUENCY_ORDER` so hopefully it's more clear that I'm filtering for period types that are shorter than yearly periods types.